File: mobile/lib/data/models/order.dart
/// Arabic labels for every order status the API can return (E3 + E4).
/// Kept top-level so both the confirmation screen header and the timeline
/// widget use the same source of truth.
String orderStatusAr(String status) {
  switch (status) {
    case 'created':
      return 'قيد التأكيد';
    case 'confirmed':
      return 'مؤكد';
    case 'preparing':
      return 'قيد التجهيز';
    case 'on_the_way':
      return 'في الطريق';
    case 'ready_for_pickup':
      return 'جاهز للاستلام';
    case 'delivered':
      return 'تم التسليم';
    case 'cancelled':
      return 'ملغى';
    case 'failed':
      return 'فشل الدفع';
    default:
      return status;
  }
}

/// Terminal (no further transitions expected) — mobile stops polling when true.
bool isTerminalOrderStatus(String status) =>
    status == 'delivered' || status == 'cancelled' || status == 'failed';

/// Shapes for the E3 orders API (mirrors backend order_schemas.py).
class OrderLineSelectionResp {
  final int groupId;
  final String groupNameAr;
  final String groupKind;
  final int optionId;
  final String optionNameAr;
  final double priceDelta;

  const OrderLineSelectionResp({
    required this.groupId,
    required this.groupNameAr,
    required this.groupKind,
    required this.optionId,
    required this.optionNameAr,
    required this.priceDelta,
  });

  factory OrderLineSelectionResp.fromJson(Map<String, dynamic> j) =>
      OrderLineSelectionResp(
        groupId: j['group_id'] as int,
        groupNameAr: j['group_name_ar'] as String,
        groupKind: j['group_kind'] as String,
        optionId: j['option_id'] as int,
        optionNameAr: j['option_name_ar'] as String,
        priceDelta: _n(j['price_delta']),
      );
}

class OrderLineResp {
  final int id;
  final int? itemId;
  final String nameAr;
  final String? imageUrl;
  final double basePrice;
  final int quantity;
  final double unitPrice;
  final double linePrice;
  final int sortOrder;
  final List<OrderLineSelectionResp> selections;

  const OrderLineResp({
    required this.id,
    this.itemId,
    required this.nameAr,
    this.imageUrl,
    required this.basePrice,
    required this.quantity,
    required this.unitPrice,
    required this.linePrice,
    required this.sortOrder,
    required this.selections,
  });

  factory OrderLineResp.fromJson(Map<String, dynamic> j) => OrderLineResp(
        id: j['id'] as int,
        itemId: j['item_id'] as int?,
        nameAr: j['name_ar'] as String,
        imageUrl: j['image_url'] as String?,
        basePrice: _n(j['base_price']),
        quantity: (j['quantity'] as num).toInt(),
        unitPrice: _n(j['unit_price']),
        linePrice: _n(j['line_price']),
        sortOrder: (j['sort_order'] as num?)?.toInt() ?? 0,
        selections: ((j['selections'] as List?) ?? const [])
            .map((e) => OrderLineSelectionResp.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class OrderResp {
  final int id;
  final String? orderNumber;
  final String status; // created | confirmed | preparing | on_the_way | ready_for_pickup | delivered | cancelled | failed
  final String fulfillmentType; // delivery | pickup
  final String? deliveryAddress;
  final String customerName;
  final String customerPhone;
  final int? customerId;
  final double subtotal;
  final double deliveryFee;
  final double discountAmount; // E5+E6 (total)
  final double pointsDiscount; // E5
  final double codeDiscount;   // E6
  final String? discountCode;  // E6 (snapshot of code as typed)
  final double total;
  final int pointsRedeemed; // E5
  final int pointsEarned;   // E5 (only nonzero once delivered)
  final String paymentMethod; // cash | apple_pay | gateway_stub
  final String? paymentReference;
  final String? notes;
  final List<OrderLineResp> lines;

  const OrderResp({
    required this.id,
    this.orderNumber,
    required this.status,
    required this.fulfillmentType,
    this.deliveryAddress,
    required this.customerName,
    required this.customerPhone,
    this.customerId,
    required this.subtotal,
    required this.deliveryFee,
    required this.discountAmount,
    required this.pointsDiscount,
    required this.codeDiscount,
    this.discountCode,
    required this.total,
    required this.pointsRedeemed,
    required this.pointsEarned,
    required this.paymentMethod,
    this.paymentReference,
    this.notes,
    required this.lines,
  });

  factory OrderResp.fromJson(Map<String, dynamic> j) => OrderResp(
        id: j['id'] as int,
        orderNumber: j['order_number'] as String?,
        status: j['status'] as String,
        fulfillmentType: j['fulfillment_type'] as String,
        deliveryAddress: j['delivery_address'] as String?,
        customerName: j['customer_name'] as String,
        customerPhone: j['customer_phone'] as String,
        customerId: j['customer_id'] as int?,
        subtotal: _n(j['subtotal']),
        deliveryFee: _n(j['delivery_fee']),
        discountAmount: _n(j['discount_amount']),
        pointsDiscount: _n(j['points_discount']),
        codeDiscount: _n(j['code_discount']),
        discountCode: j['discount_code'] as String?,
        total: _n(j['total']),
        pointsRedeemed: (j['points_redeemed'] as num?)?.toInt() ?? 0,
        pointsEarned: (j['points_earned'] as num?)?.toInt() ?? 0,
        paymentMethod: j['payment_method'] as String,
        paymentReference: j['payment_reference'] as String?,
        notes: j['notes'] as String?,
        lines: ((j['lines'] as List?) ?? const [])
            .map((e) => OrderLineResp.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// Result of POST /orders. Payment goes into a compact wrapper.
class OrderCreateResp {
  final OrderResp order;
  final String paymentStatus;   // confirmed | redirect | failed
  final String? redirectUrl;
  final String? reference;
  final String? message;

  const OrderCreateResp({
    required this.order,
    required this.paymentStatus,
    this.redirectUrl,
    this.reference,
    this.message,
  });

  factory OrderCreateResp.fromJson(Map<String, dynamic> j) {
    final pay = (j['payment'] as Map<String, dynamic>?) ?? const {};
    return OrderCreateResp(
      order: OrderResp.fromJson(j['order'] as Map<String, dynamic>),
      paymentStatus: (pay['status'] as String?) ?? 'failed',
      redirectUrl: pay['redirect_url'] as String?,
      reference: pay['reference'] as String?,
      message: pay['message'] as String?,
    );
  }
}

double _n(dynamic v) {
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v) ?? 0.0;
  return 0.0;
}

File: backend/app/models/order.py
"""Order + OrderLine + OrderLineSelection — the first server-side representation
of a purchase (E1 and E2 were client-side only).

Everything price / name / image related is SNAPSHOTTED at order-creation time
so future admin edits to the source items don't rewrite past orders.

`order_number` (human-facing, e.g. AT-2026-000123) is generated in the
`before_insert` observer using the assigned PK. Registered in app/observers.py.
"""
from datetime import datetime, timezone
from decimal import Decimal
from enum import Enum

from app.extensions import db


class OrderStatus(str, Enum):
    # payment-flow states (E3)
    created = "created"                    # order placed, awaiting gateway callback
    confirmed = "confirmed"                # cash-immediate or gateway-callback success
    failed = "failed"                      # gateway declined / errored
    # workflow states (E4 — set by admin)
    preparing = "preparing"                # قيد التجهيز
    on_the_way = "on_the_way"              # في الطريق (delivery only)
    ready_for_pickup = "ready_for_pickup"  # جاهز للاستلام (pickup only)
    delivered = "delivered"                # تم التسليم / تم الاستلام
    # terminal-cancel
    cancelled = "cancelled"


# Terminal statuses — nothing follows.
_TERMINAL_STATUSES: set[OrderStatus] = {
    OrderStatus.delivered,
    OrderStatus.cancelled,
    OrderStatus.failed,
}


class FulfillmentType(str, Enum):
    delivery = "delivery"
    pickup = "pickup"


class PaymentMethod(str, Enum):
    cash = "cash"
    apple_pay = "apple_pay"
    gateway_stub = "gateway_stub"


class Order(db.Model):
    __tablename__ = "orders"

    id = db.Column(db.Integer, primary_key=True)
    order_number = db.Column(db.String(32), unique=True, nullable=True, index=True)

    status = db.Column(
        db.Enum(OrderStatus, name="order_status"),
        nullable=False,
        default=OrderStatus.created,
    )

    fulfillment_type = db.Column(
        db.Enum(FulfillmentType, name="order_fulfillment_type"), nullable=False
    )
    delivery_address = db.Column(db.String(500), nullable=True)

    customer_name = db.Column(db.String(120), nullable=False)
    customer_phone = db.Column(db.String(40), nullable=False)
    # E5 — set at order-create by upserting the phone. Nullable so old E1–E4 orders (no customer row) stay valid.
    customer_id = db.Column(
        db.Integer, db.ForeignKey("customers.id", ondelete="SET NULL"), nullable=True, index=True
    )

    subtotal = db.Column(db.Numeric(10, 2), nullable=False, default=Decimal("0"))
    delivery_fee = db.Column(db.Numeric(10, 2), nullable=False, default=Decimal("0"))
    discount_amount = db.Column(db.Numeric(10, 2), nullable=False, default=Decimal("0"), server_default="0")
    total = db.Column(db.Numeric(10, 2), nullable=False, default=Decimal("0"))

    # E5 loyalty
    points_redeemed = db.Column(db.Integer, nullable=False, default=0, server_default="0")
    points_earned = db.Column(db.Integer, nullable=False, default=0, server_default="0")
    points_discount = db.Column(
        db.Numeric(10, 2), nullable=False, default=Decimal("0"), server_default="0"
    )

    # E6 discount code
    code_discount = db.Column(
        db.Numeric(10, 2), nullable=False, default=Decimal("0"), server_default="0"
    )
    discount_code_id = db.Column(
        db.Integer, db.ForeignKey("discount_codes.id", ondelete="SET NULL"), nullable=True
    )
    # Snapshot so a later-deleted code still shows on the receipt.
    discount_code_snapshot = db.Column(db.String(40), nullable=True)

    payment_method = db.Column(
        db.Enum(PaymentMethod, name="order_payment_method"), nullable=False
    )
    payment_reference = db.Column(db.String(120), nullable=True)

    notes = db.Column(db.Text, nullable=True)

    created_at = db.Column(db.DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    confirmed_at = db.Column(db.DateTime(timezone=True), nullable=True)
    # E4 — set the first time an admin opens this order's detail page. Drives
    # the "جديد" badge + row highlight on the orders list.
    admin_seen_at = db.Column(db.DateTime(timezone=True), nullable=True)
    updated_at = db.Column(
        db.DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    lines = db.relationship(
        "OrderLine",
        back_populates="order",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="OrderLine.sort_order",
    )
    customer = db.relationship("Customer")

    def __repr__(self) -> str:  # pragma: no cover
        return f"<Order #{self.id} {self.order_number} {self.status.value} total={self.total}>"

    # ---- E4 workflow helpers ----

    def next_valid_statuses(self) -> list[OrderStatus]:
        """Statuses the admin may transition to from the current one.

        Enforces the "logical order only" rule from US4.2. The admin UI
        renders one button per entry here; the transition endpoint validates
        against the same list (belt-and-braces).
        """
        s = self.status
        if s == OrderStatus.confirmed:
            return [OrderStatus.preparing, OrderStatus.cancelled]
        if s == OrderStatus.preparing:
            if self.fulfillment_type == FulfillmentType.delivery:
                return [OrderStatus.on_the_way, OrderStatus.cancelled]
            return [OrderStatus.ready_for_pickup, OrderStatus.cancelled]
        if s == OrderStatus.on_the_way or s == OrderStatus.ready_for_pickup:
            return [OrderStatus.delivered, OrderStatus.cancelled]
        # created/failed/delivered/cancelled: nothing admin-transitionable
        return []

    def is_terminal(self) -> bool:
        return self.status in _TERMINAL_STATUSES


class OrderLine(db.Model):
    __tablename__ = "order_lines"

    id = db.Column(db.Integer, primary_key=True)
    order_id = db.Column(
        db.Integer, db.ForeignKey("orders.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # SET NULL on delete so an item deletion doesn't erase order history.
    item_id = db.Column(db.Integer, db.ForeignKey("items.id", ondelete="SET NULL"), nullable=True, index=True)

    name_ar_snapshot = db.Column(db.String(160), nullable=False)
    image_path_snapshot = db.Column(db.String(255), nullable=True)
    base_price_snapshot = db.Column(db.Numeric(10, 2), nullable=False)

    quantity = db.Column(db.Integer, nullable=False, default=1)
    unit_price = db.Column(db.Numeric(10, 2), nullable=False)  # base + all deltas
    line_price = db.Column(db.Numeric(10, 2), nullable=False)  # unit_price * quantity

    sort_order = db.Column(db.Integer, nullable=False, default=0)

    order = db.relationship("Order", back_populates="lines")
    selections = db.relationship(
        "OrderLineSelection",
        back_populates="line",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )


class OrderLineSelection(db.Model):
    __tablename__ = "order_line_selections"

    id = db.Column(db.Integer, primary_key=True)
    order_line_id = db.Column(
        db.Integer,
        db.ForeignKey("order_lines.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    group_id_snapshot = db.Column(db.Integer, nullable=False)
    group_name_ar_snapshot = db.Column(db.String(120), nullable=False)
    group_kind_snapshot = db.Column(db.String(20), nullable=False)  # 'variant'|'size'|'remove'|'add'
    option_id_snapshot = db.Column(db.Integer, nullable=False)
    option_name_ar_snapshot = db.Column(db.String(120), nullable=False)
    price_delta_snapshot = db.Column(db.Numeric(10, 2), nullable=False, default=Decimal("0"))

    line = db.relationship("OrderLine", back_populates="selections")
